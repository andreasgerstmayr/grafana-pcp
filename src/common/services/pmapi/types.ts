import { DataSourceInstanceSettings } from '@grafana/data';
import { InstanceName, Labels, MetricName, Semantics } from 'common/types/pcp';

export interface PmApiConfig {
    dsInstanceSettings: DataSourceInstanceSettings;
    timeoutMs: number;
}

export type InstanceId = number;

export interface Context {
    context: number;
    labels: Labels;
}

export interface MetricResponse {
    metrics: Metadata[];
}

export interface Metadata {
    name: MetricName;
    indom?: string;
    type: string;
    sem: Semantics;
    series: string;
    units: string;
    labels: Labels;
    'text-help': string;
    'text-oneline': string;
}

export interface IndomResponse {
    instances: Indom[];
    labels: Labels;
}

export interface Indom {
    name: InstanceName;
    instance: InstanceId;
    labels: Labels;
}

export interface FetchResponse {
    timestamp: number;
    values: MetricInstanceValues[];
}

export interface MetricInstanceValues {
    name: MetricName;
    instances: InstanceValue[];
}

export interface InstanceValue {
    instance: InstanceId | null;
    value: number | string;
}

export interface StoreResponse {
    success: boolean;
}

export interface DeriveResponse {
    success: boolean;
}

export interface ChildrenResponse {
    leaf: string[];
    nonleaf: string[];
}

export class MetricNotFoundError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Cannot find metric ${metric}. Please check if the PMDA is enabled.`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class NoIndomError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Metric ${metric} has no instance domain.`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class MetricSemanticError extends Error {
    constructor(readonly expr: string, message?: string) {
        super(message ?? `Semantic error in '${expr}' definition.`);
        this.expr = expr;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class MetricSyntaxError extends Error {
    constructor(readonly expr: string, message?: string) {
        super(message ?? `Syntax error in '${expr}' definition.`);
        this.expr = expr;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class DuplicateDerivedMetricNameError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Duplicate derived metric name ${metric}`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class PermissionError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Insufficient permissions to store metric ${metric}. Please check the PMDA configuration.`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
