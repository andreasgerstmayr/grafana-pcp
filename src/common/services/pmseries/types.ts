import { DataSourceInstanceSettings } from '@grafana/data';
import { InstanceName, Labels, Semantics } from 'common/types/pcp';

export interface PmSeriesApiConfig {
    dsInstanceSettings: DataSourceInstanceSettings;
    timeoutMs: number;
    baseUrl: string;
}

export type SeriesId = string;
export type SeriesInstanceId = string;

export interface SeriesNoRecordResponse {
    success: boolean;
}

export interface SeriesDescQueryParams {
    series: string[]; // list of series identifiers
}
export interface SeriesDescItemResponse {
    series: string;
    source: string;
    pmid: string;
    indom: string;
    semantics: Semantics;
    type: string;
    units: string;
}
export type SeriesDescResponse = SeriesDescItemResponse[];

// this one has no related 'maybe' response type
export interface SeriesPingResponse {
    success: boolean;
}

export interface SeriesQueryQueryParams {
    expr: string; // Query string in [pmseries](https://pcp.io/man/man1/pmseries.1.html) format
}
export interface SeriesQueryItemResponse {
    series: string;
    instance: string;
    timestamp: number;
    value: string;
}
export type SeriesQueryResponse = string[] | SeriesQueryItemResponse[];

export interface SeriesInstancesQueryParams {
    /** Comma-separated list of series identifiers */
    series?: string[];
    /** Glob pattern string to match on all labels */
    match?: string;
}
export interface SeriesInstancesItemResponse {
    series: SeriesId;
    source: string;
    instance: SeriesInstanceId;
    id: number;
    name: InstanceName;
}
export type SeriesInstancesResponse = SeriesInstancesItemResponse[];

export interface SeriesLabelsQueryParams {
    series?: string[]; // Comma-separated list of series identifiers
    match?: string; // Glob pattern string to match on all labels
    name?: string; // Find all known label values for given name
    names?: string[]; // Comma-separated list of label names
}
export interface SeriesLabelsItemResponse {
    series: string;
    labels: Labels;
}
export interface SeriesLabelsLabelValuesItemResponse {
    [key: string]: Array<string | number | boolean>;
}
export type SeriesLabelsResponse = string[] | SeriesLabelsItemResponse[] | SeriesLabelsLabelValuesItemResponse;

export interface SeriesMetricsQueryParams {
    series?: string[];
    match?: string;
}
export interface SeriesMetricsItemResponse {
    series: string;
    name: string;
}
export type SeriesMetricsResponse = string[] | SeriesMetricsItemResponse[];

export interface SeriesValuesQueryParams {
    series?: string[];
    samples?: number;
    interval?: string;
    start?: string;
    finish?: string;
    offset?: string;
    align?: string;
    zone?: string;
}
export interface SeriesValuesItemResponse {
    series: string;
    instance?: string;
    /** unix timestamp in milliseconds */
    timestamp: number;
    value: string;
}
export type SeriesValuesResponse = SeriesValuesItemResponse[];
