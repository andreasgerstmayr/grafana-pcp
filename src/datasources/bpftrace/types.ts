import { Optional } from 'utility-types';
import { MinimalPmapiQuery, PmapiOptions, PmapiQuery } from '../../datasources/lib/pmapi/types';
import { TargetFormat } from '../../datasources/lib/types';
import { Script } from './script';

export interface BPFtraceOptions extends PmapiOptions {}

export interface BPFtraceQuery extends MinimalPmapiQuery {}

export const defaultBPFtraceQuery: BPFtraceQuery & Optional<PmapiQuery, 'url' | 'hostspec'> = {
    refId: 'A',
    expr: '',
    format: TargetFormat.TimeSeries,
    options: {
        rateConversation: true,
        timeUtilizationConversation: true,
    },
};

export interface BPFtraceTargetData {
    script: Script;
}
