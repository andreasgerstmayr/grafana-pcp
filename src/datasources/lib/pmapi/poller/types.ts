import { Context, Indom, InstanceValue, Metadata } from 'common/services/pmapi/types';
import { Labels } from 'common/types/pcp';
import { Dict, RequiredField } from 'common/types/utils';
import { Target, TemplatedPmapiQuery } from '../types';

export interface Metric {
    metadata: Metadata;
    instanceDomain?: {
        instances: Dict<string, Indom>;
        labels: Labels;
    };
    values: InstanceValuesSnapshot[];
}

export interface InstanceValuesSnapshot {
    timestampMs: number;
    values: InstanceValue[];
}

export enum EndpointState {
    /** new entered endpoint, no context available */
    PENDING,
    /** context available */
    CONNECTED,
}

/**
 * single endpoint, identified by url and hostspec
 * each url/hostspec has a different context
 * each url/hostspec can have different metrics (and values)
 */
export interface Endpoint {
    state: EndpointState;
    url: string;
    hostspec: string;
    metrics: Metric[];
    targets: Target[];
    additionalMetricsToPoll: Array<{ name: string; callback: (values: InstanceValue[]) => void }>;
    errors: any[];

    /** context, will be created at next poll */
    context?: Context;
    /** backfilling with redis */
    hasRedis?: boolean;
}

export type EndpointWithCtx = RequiredField<Endpoint, 'context'>;

export interface QueryResult {
    endpoint: EndpointWithCtx;
    query: TemplatedPmapiQuery;
    metrics: Metric[];
}
