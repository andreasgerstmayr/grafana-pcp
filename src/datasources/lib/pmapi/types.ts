import { DataQuery, DataSourceJsonData } from '@grafana/data';
import { RequiredField } from 'common/types/utils';
import { TargetFormat } from '../types';

export interface PmapiOptions extends DataSourceJsonData {
    hostspec?: string;
    retentionTime?: string;
}

export interface PmapiDefaultOptions {
    hostspec: string;
    retentionTime: string;
}

export interface PmapiQuery extends DataQuery {
    expr: string;
    format: TargetFormat;
    legendFormat?: string;

    url?: string;
    hostspec?: string;
}

export type TemplatedPmapiQuery = RequiredField<PmapiQuery, 'url' | 'hostspec'>;

export enum TargetState {
    /** newly entered target or target with error (trying again) */
    PENDING,
    /** metrics exists and metadata available */
    METRICS_AVAILABLE,
    /** fatal error, will not try again */
    ERROR,
}

/**
 * Represents a target of a Grafana panel, which will be polled in the background
 * extends the Query (as provided by Grafana) with additional information:
 * - vector+bpftrace: errors occured while polling in the background, lastActive, ...
 * - vector: isDerivedMetric
 * - bpftrace: script
 * is persisted during multiple queries (key = targetId)
 */
export interface Target<T = any> {
    targetId: string;
    state: TargetState;
    query: TemplatedPmapiQuery;
    /** valid PCP metric names (can be a derived metric, e.g. derived_xxx) */
    metricNames: string[];
    errors: any[];
    lastActiveMs: number;
    custom?: T;
}
