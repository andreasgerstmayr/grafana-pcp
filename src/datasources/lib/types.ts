export enum TargetFormat {
    TimeSeries = 'time_series',
    Heatmap = 'heatmap',
    /** vector only */
    MetricsTable = 'metrics_table',
    /** bpftrace only */
    CsvTable = 'csv_table',
    /** bpftrace only */
    FlameGraph = 'flamegraph',
}
