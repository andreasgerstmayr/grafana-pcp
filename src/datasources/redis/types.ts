import { DataQuery, DataSourceJsonData } from '@grafana/data';
import { TargetFormat } from '../../datasources/lib/types';

export interface RedisOptions extends DataSourceJsonData {}

export interface RedisQuery extends DataQuery {
    expr: string;
    format: TargetFormat;
    legendFormat?: string;
}

export const defaultRedisQuery: Partial<RedisQuery> = {
    expr: '',
    format: TargetFormat.TimeSeries,
};
