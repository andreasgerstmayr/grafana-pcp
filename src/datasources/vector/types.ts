import { PmapiOptions, PmapiQuery, TargetFormat } from 'datasources/lib/pmapi/types';

export interface VectorOptions extends PmapiOptions {}

export interface VectorQuery extends PmapiQuery {}

export const defaultVectorQuery: Partial<VectorQuery> = {
    expr: '',
    format: TargetFormat.TimeSeries,
};

export interface VectorTargetData {
    isDerivedMetric: boolean;
}
