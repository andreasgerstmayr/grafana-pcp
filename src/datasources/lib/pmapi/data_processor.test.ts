import { grafana, pcp, poller } from '../specs/fixtures';
import { TargetFormat } from '../types';
import { processQueries } from './data_processor';

describe('data processor', () => {
    it('should create a dataframe and handle missing frames and backward counters', () => {
        const target = poller.target({ query: { expr: 'disk.dev.read' } });
        const dataQueryRequest = grafana.dataQueryRequest([target.query]); // request data between 10-20s

        const values = [
            {
                timestampMs: 7000, // out of range
                values: [
                    { instance: 0, value: 7 },
                    { instance: 1, value: 5 },
                ],
            },
            {
                timestampMs: 8000, // request 1 value more to fill the graph
                values: [
                    { instance: 0, value: 8 },
                    { instance: 1, value: 5 },
                ],
            },
            {
                timestampMs: 9000, // request 1 value more because of counter metric
                values: [
                    { instance: 0, value: 9 },
                    { instance: 1, value: 5 },
                ],
            },
            {
                timestampMs: 10000,
                values: [
                    { instance: 0, value: 10 },
                    { instance: 1, value: 6 },
                ],
            },
            {
                timestampMs: 11000,
                values: [
                    { instance: 0, value: 13 },
                    { instance: 1, value: 7 },
                ],
            },
            {
                timestampMs: 12000,
                values: [
                    { instance: 0, value: 14 },
                    // instance 1 missing
                ],
            },
            {
                timestampMs: 13000,
                values: [
                    { instance: 0, value: 15 },
                    { instance: 1, value: 8 },
                ],
            },
            {
                timestampMs: 14000,
                values: [
                    { instance: 0, value: 16 },
                    { instance: 1, value: 9 },
                ],
            },
            {
                timestampMs: 15000,
                values: [
                    { instance: 0, value: 18 },
                    { instance: 1, value: 8 }, // counter went backwards
                ],
            },
            {
                timestampMs: 16000,
                values: [
                    { instance: 0, value: 19 },
                    { instance: 1, value: 9 },
                ],
            },
            {
                timestampMs: 17000,
                values: [
                    { instance: 0, value: 20 },
                    { instance: 1, value: 10 },
                ],
            },
            {
                timestampMs: 21000, // request 1 value more to fill the graph
                values: [
                    { instance: 0, value: 21 },
                    { instance: 1, value: 11 },
                ],
            },
            {
                timestampMs: 22000, // out of range
                values: [
                    { instance: 0, value: 22 },
                    { instance: 1, value: 12 },
                ],
            },
        ];
        const metric = {
            ...pcp.metrics['disk.dev.read'],
            values,
        };
        const endpoint = poller.endpoint({ metrics: [metric], targets: [target] });

        const result = processQueries(dataQueryRequest, [{ endpoint, query: target.query, metrics: [metric] }], 1);
        expect({ fields: result[0].fields }).toMatchInlineSnapshot(
            {
                fields: [{}, { config: { custom: expect.anything() } }, { config: { custom: expect.anything() } }],
            },
            `
            Object {
              "fields": Array [
                Object {
                  "config": Object {},
                  "name": "Time",
                  "type": "time",
                  "values": Array [
                    8000,
                    9000,
                    10000,
                    11000,
                    12000,
                    13000,
                    14000,
                    15000,
                    16000,
                    17000,
                    21000,
                  ],
                },
                Object {
                  "config": Object {
                    "custom": Anything,
                    "displayNameFromDS": "",
                  },
                  "labels": Object {
                    "agent": "linux",
                    "device_type": "block",
                    "domainname": "localdomain",
                    "hostname": "dev",
                    "indom_name": "per disk",
                    "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                  },
                  "name": "disk.dev.read[nvme0n1]",
                  "type": "number",
                  "values": Array [
                    null,
                    1,
                    1,
                    3,
                    1,
                    1,
                    1,
                    2,
                    1,
                    1,
                    0.25,
                  ],
                },
                Object {
                  "config": Object {
                    "custom": Anything,
                    "displayNameFromDS": "",
                  },
                  "labels": Object {
                    "agent": "linux",
                    "device_type": "block",
                    "domainname": "localdomain",
                    "hostname": "dev",
                    "indom_name": "per disk",
                    "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                  },
                  "name": "disk.dev.read[sda]",
                  "type": "number",
                  "values": Array [
                    null,
                    0,
                    1,
                    1,
                    null,
                    null,
                    1,
                    null,
                    1,
                    1,
                    0.25,
                  ],
                },
              ],
            }
        `
        );
    });

    it('should process a metrics table', () => {
        const targetA = poller.target({
            query: { expr: 'some.string.A', refId: 'A', format: TargetFormat.MetricsTable },
        });
        const targetB = poller.target({
            query: { expr: 'some.string.B', refId: 'B', format: TargetFormat.MetricsTable },
        });

        const metricA = {
            metadata: pcp.metadata({
                name: 'some.string.A',
                type: 'string',
                units: 'none',
            }),
            values: [
                {
                    timestampMs: 10000,
                    values: [
                        { instance: 0, value: 'A/0/10000' },
                        { instance: 1, value: 'A/1/10000' },
                    ],
                },
                {
                    timestampMs: 11000,
                    values: [
                        { instance: 0, value: 'A/0/11000' },
                        { instance: 1, value: 'A/1/11000' },
                    ],
                },
            ],
            instanceDomain: {
                instances: {
                    0: { name: 'Inst 0', instance: 0, labels: {} },
                    1: { name: 'Inst 1', instance: 1, labels: {} },
                },
                labels: {},
            },
        };
        const metricB = {
            metadata: pcp.metadata({
                name: 'some.string.B',
                type: 'string',
                units: 'none',
            }),
            values: [
                {
                    timestampMs: 10000,
                    values: [
                        { instance: 0, value: 'B/0/10000' },
                        { instance: 1, value: 'B/1/10000' },
                    ],
                },
                {
                    timestampMs: 11000,
                    values: [
                        { instance: 0, value: 'B/0/11000' },
                        { instance: 1, value: 'B/1/11000' },
                    ],
                },
            ],
            instanceDomain: {
                instances: {
                    0: { name: 'Inst 0', instance: 0, labels: {} },
                    1: { name: 'Inst 1', instance: 1, labels: {} },
                },
                labels: {},
            },
        };

        const endpoint = poller.endpoint({ metrics: [metricA, metricB], targets: [targetA, targetB] });
        const dataQueryRequest = grafana.dataQueryRequest([targetA.query, targetB.query]);

        const result = processQueries(
            dataQueryRequest,
            [
                { endpoint, query: targetA.query, metrics: [metricA] },
                { endpoint, query: targetB.query, metrics: [metricB] },
            ],
            1
        );
        expect({ fields: result[0].fields }).toMatchInlineSnapshot(
            { fields: [{}, { config: { custom: expect.anything() } }, { config: { custom: expect.anything() } }] },
            `
            Object {
              "fields": Array [
                Object {
                  "config": Object {},
                  "name": "instance",
                  "type": "string",
                  "values": Array [
                    "Inst 0",
                    "Inst 1",
                  ],
                },
                Object {
                  "config": Object {
                    "custom": Anything,
                    "displayNameFromDS": "A",
                  },
                  "labels": Object {
                    "agent": "linux",
                    "hostname": "host1",
                  },
                  "name": "some.string.A",
                  "type": "string",
                  "values": Array [
                    "A/0/11000",
                    "A/1/11000",
                  ],
                },
                Object {
                  "config": Object {
                    "custom": Anything,
                    "displayNameFromDS": "B",
                  },
                  "labels": Object {
                    "agent": "linux",
                    "hostname": "host1",
                  },
                  "name": "some.string.B",
                  "type": "string",
                  "values": Array [
                    "B/0/11000",
                    "B/1/11000",
                  ],
                },
              ],
            }
        `
        );
    });

    it('should process a CSV table', () => {
        const target = poller.target({
            query: { expr: 'some.string', refId: 'A', format: TargetFormat.CsvTable },
        });
        const metric = {
            metadata: pcp.metadata({
                name: 'some.string',
                type: 'string',
                units: 'none',
            }),
            values: [
                {
                    timestampMs: 10000,
                    values: [{ instance: null, value: 'a,b,c' }],
                },
                {
                    timestampMs: 11000,
                    values: [
                        {
                            instance: null,
                            value:
                                'col1,col2,col3\n' +
                                'row1 col1,row1 col2,row1 col3\n' +
                                'row2 col1,row2 col2,row2 col3',
                        },
                    ],
                },
            ],
        };

        const endpoint = poller.endpoint({ metrics: [metric], targets: [target] });
        const dataQueryRequest = grafana.dataQueryRequest([target.query]);
        const result = processQueries(dataQueryRequest, [{ endpoint, query: target.query, metrics: [metric] }], 1);
        expect(result[0].fields).toMatchInlineSnapshot(`
            Array [
              Object {
                "config": Object {},
                "name": "col1",
                "type": "string",
                "values": Array [
                  "row1 col1",
                  "row2 col1",
                ],
              },
              Object {
                "config": Object {},
                "name": "col2",
                "type": "string",
                "values": Array [
                  "row1 col2",
                  "row2 col2",
                ],
              },
              Object {
                "config": Object {},
                "name": "col3",
                "type": "string",
                "values": Array [
                  "row1 col3",
                  "row2 col3",
                ],
              },
            ]
        `);
    });
});
