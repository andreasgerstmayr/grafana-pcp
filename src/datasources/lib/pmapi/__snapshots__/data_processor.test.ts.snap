// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data processor should create a dataframe and handle missing frames and backward counters 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "Time",
        "type": "time",
        "values": Array [
          8000,
          9000,
          10000,
          11000,
          12000,
          13000,
          14000,
          15000,
          16000,
          17000,
          21000,
        ],
      },
      Object {
        "config": Object {
          "custom": Object {
            "instance": Object {
              "instance": 0,
              "labels": Object {
                "domainname": "localdomain",
                "hostname": "dev",
                "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
              },
              "name": "nvme0n1",
            },
            "instanceId": 0,
          },
          "displayName": "",
          "unit": undefined,
        },
        "labels": Object {
          "agent": "linux",
          "device_type": "block",
          "domainname": "localdomain",
          "hostname": "dev",
          "indom_name": "per disk",
          "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
        },
        "name": "disk.dev.read[nvme0n1]",
        "type": "number",
        "values": Array [
          null,
          1,
          1,
          3,
          1,
          1,
          1,
          2,
          1,
          1,
          0.25,
        ],
      },
      Object {
        "config": Object {
          "custom": Object {
            "instance": Object {
              "instance": 1,
              "labels": Object {
                "domainname": "localdomain",
                "hostname": "dev",
                "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
              },
              "name": "sda",
            },
            "instanceId": 1,
          },
          "displayName": "",
          "unit": undefined,
        },
        "labels": Object {
          "agent": "linux",
          "device_type": "block",
          "domainname": "localdomain",
          "hostname": "dev",
          "indom_name": "per disk",
          "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
        },
        "name": "disk.dev.read[sda]",
        "type": "number",
        "values": Array [
          null,
          0,
          1,
          1,
          null,
          null,
          1,
          null,
          1,
          1,
          0.25,
        ],
      },
    ],
    "meta": Object {
      "custom": Object {
        "context": Object {
          "context": 123,
          "labels": Object {},
        },
        "metric": Object {
          "instanceDomain": Object {
            "instances": Object {
              "0": Object {
                "instance": 0,
                "labels": Object {
                  "domainname": "localdomain",
                  "hostname": "dev",
                  "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                },
                "name": "nvme0n1",
              },
              "1": Object {
                "instance": 1,
                "labels": Object {
                  "domainname": "localdomain",
                  "hostname": "dev",
                  "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                },
                "name": "sda",
              },
            },
            "labels": Object {
              "device_type": "block",
              "domainname": "localdomain",
              "hostname": "dev",
              "indom_name": "per disk",
              "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
            },
          },
          "metadata": Object {
            "indom": "60.1",
            "labels": Object {
              "agent": "linux",
              "device_type": "block",
              "domainname": "localdomain",
              "hostname": "dev",
              "indom_name": "per disk",
              "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
            },
            "name": "disk.dev.read",
            "sem": "counter",
            "series": "f87250c4ea0e5eca8ff2ca3b3044ba1a6c91a3d9",
            "text-help": "Cumulative number of disk read operations since system boot time (subject
to counter wrap).",
            "text-oneline": "per-disk read operations",
            "type": "u64",
            "units": "count",
          },
          "values": Array [
            Object {
              "timestampMs": 7000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 7,
                },
                Object {
                  "instance": 1,
                  "value": 5,
                },
              ],
            },
            Object {
              "timestampMs": 8000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 8,
                },
                Object {
                  "instance": 1,
                  "value": 5,
                },
              ],
            },
            Object {
              "timestampMs": 9000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 9,
                },
                Object {
                  "instance": 1,
                  "value": 5,
                },
              ],
            },
            Object {
              "timestampMs": 10000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 10,
                },
                Object {
                  "instance": 1,
                  "value": 6,
                },
              ],
            },
            Object {
              "timestampMs": 11000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 13,
                },
                Object {
                  "instance": 1,
                  "value": 7,
                },
              ],
            },
            Object {
              "timestampMs": 12000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 14,
                },
              ],
            },
            Object {
              "timestampMs": 13000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 15,
                },
                Object {
                  "instance": 1,
                  "value": 8,
                },
              ],
            },
            Object {
              "timestampMs": 14000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 16,
                },
                Object {
                  "instance": 1,
                  "value": 9,
                },
              ],
            },
            Object {
              "timestampMs": 15000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 18,
                },
                Object {
                  "instance": 1,
                  "value": 8,
                },
              ],
            },
            Object {
              "timestampMs": 16000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 19,
                },
                Object {
                  "instance": 1,
                  "value": 9,
                },
              ],
            },
            Object {
              "timestampMs": 17000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 20,
                },
                Object {
                  "instance": 1,
                  "value": 10,
                },
              ],
            },
            Object {
              "timestampMs": 21000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 21,
                },
                Object {
                  "instance": 1,
                  "value": 11,
                },
              ],
            },
            Object {
              "timestampMs": 22000,
              "values": Array [
                Object {
                  "instance": 0,
                  "value": 22,
                },
                Object {
                  "instance": 1,
                  "value": 12,
                },
              ],
            },
          ],
        },
        "query": Object {
          "expr": "disk.dev.read",
          "format": "time_series",
          "hostspec": "",
          "refId": "A",
          "url": "",
        },
      },
    },
    "name": undefined,
    "refId": undefined,
  },
]
`;

exports[`data processor should process a CSV table 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "col1",
        "type": "other",
        "values": Array [
          "row1 col1",
          "row2 col1",
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "col2",
        "type": "other",
        "values": Array [
          "row1 col2",
          "row2 col2",
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "col3",
        "type": "other",
        "values": Array [
          "row1 col3",
          "row2 col3",
        ],
      },
    ],
    "meta": undefined,
    "name": undefined,
    "refId": undefined,
  },
]
`;

exports[`data processor should process a metrics table 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "instance",
        "type": "string",
        "values": Array [
          "Inst 0",
          "Inst 1",
        ],
      },
      Object {
        "config": Object {
          "custom": Object {
            "instance": undefined,
            "instanceId": null,
          },
          "displayName": "A",
          "unit": undefined,
        },
        "labels": Object {
          "agent": "linux",
          "hostname": "host1",
        },
        "name": "some.string.A",
        "type": "string",
        "values": Array [
          "A/0/11000",
          "A/1/11000",
        ],
      },
      Object {
        "config": Object {
          "custom": Object {
            "instance": undefined,
            "instanceId": null,
          },
          "displayName": "B",
          "unit": undefined,
        },
        "labels": Object {
          "agent": "linux",
          "hostname": "host1",
        },
        "name": "some.string.B",
        "type": "string",
        "values": Array [
          "B/0/11000",
          "B/1/11000",
        ],
      },
    ],
    "meta": undefined,
    "name": undefined,
    "refId": undefined,
  },
]
`;
